### ISSUE TITLE
Projects with same app name seem to not get full snippet support (Updated)

### ISSUE THREAD

**Issue by dshimo** [OpenLiberty/liberty-tools-vscode#259]
Choosing a parent dir with multiple projects as the main folder seems to mess with our language server snippet suggestions. Instead, if you open each project's root folder through `Add folder to Workspace...`, normal behavior resumes.

My project structure has a dir named 'starters' which contain the Maven and Gradle starter apps. Each app gets recognized by the dashboard correctly. Liberty actions work.

![image](https://github.com/OpenLiberty/liberty-tools-vscode/assets/8934136/bf7e27b0-3a74-4f9f-96c5-c829167b635f)

It seems MicroProfile snippets only appear for one project. In this case, it only appears on my Gradle Java files.

![image](https://github.com/OpenLiberty/liberty-tools-vscode/assets/8934136/ad01cc24-de01-4e05-8597-174859d9ecbf)

![image](https://github.com/OpenLiberty/liberty-tools-vscode/assets/8934136/6ec103fa-90fb-4cce-b15d-19d521a121ea)

Attempting to fetch `rest_*` snippets in Maven yield irrelevant answers, and attempting `rest_*` in the Gradle project gave me only `rest_get`.

![image](https://github.com/OpenLiberty/liberty-tools-vscode/assets/8934136/372214d0-e431-491c-8683-a562b6dd5342)

This final screenshot is the expected snippet list for `rest_*`.  Here, only the Maven project is open in the workspace. 

![image](https://github.com/OpenLiberty/liberty-tools-vscode/assets/8934136/53d9c965-8f64-4216-a688-b6585293e37a)

**Comment by dshimo** [comment]
Testing on Mac, it seems as if your application names are the same in the Liberty dashboard, then snippets will fail on multi-projects.

Here, both the Maven and Gradle apps were added to the workspace. Snippets fail.
<img width="585" alt="image" src="https://github.com/OpenLiberty/liberty-tools-vscode/assets/8934136/75999d28-1cfe-436e-84ee-b5aa697d52fc">

In this screenshot, Maven was added after Gradle. `rest_class` is still not found.
<img width="587" alt="image" src="https://github.com/OpenLiberty/liberty-tools-vscode/assets/8934136/666f3e71-c0be-423d-8d0a-cc52979e36ab">

And lastly, with newly generated apps with different app names, `rest_class` returns. Here, I even opened the parent dir 'starters'
<img width="574" alt="image" src="https://github.com/OpenLiberty/liberty-tools-vscode/assets/8934136/c48e9e52-9422-421a-a97c-d029482da0f3">

**Comment by TrevCraw** [comment]
> So this only occurs when it is the same app name in more than one project? If so, that does not seem like it would be very common in the real world?

Yes, it appears the app names must be the same to hit this bug. I agree, I don't think this is a likely scenario for users. I don't think this should be a stop ship.

**Comment by anusreelakshmi934** [comment]
Receiving the following error when opening a Maven project folder:
The Language Support for Java server crashed 5 times in the last 3 minutes. The server will not be restarted.
Tried Java: Clean Java Language Server Workspace, but still error persists.
![image](https://github.com/OpenLiberty/liberty-tools-vscode/assets/141303561/2ecef397-5bce-4a32-8942-1857d910ddde)

**Comment by anusreelakshmi934** [comment]
When we open a directory containing Maven and Gradle starter apps with the same app name, one of the files is recognized as a non-project file, triggering a diagnostic message like "filename.java is a non-project file, only syntax errors are reported." In this file, snippet suggestions is not provided. However, in the other file within the same directory, suggestions are available.

Alternatively, when we open each project's root folder through "Add folder to Workspace," no diagnostic appears even if the app names are the same, and snippet suggestions are provided.
Attaching the screenshot below:
<img width="1156" alt="image" src="https://github.com/OpenLiberty/liberty-tools-vscode/assets/141303561/4bc471e9-d4f6-4365-ac2a-48a8e1f7c691">

**Comment by anusreelakshmi934** [comment]
Based on my understanding, the issue seems to be that Visual Studio Code treats all Java projects within the same workspace as a single project, causing interference between them. Despite attempting various solutions such as:
cleaning the Java language server workspace, adding folders to the Java source path, and renaming packages, setting sourcePaths in the the settings.json file to src, still the problem persists.

Through my investigation, it appears that the only effective way to avoid the warning and receive snippet suggestions is to individually open each project's root folder by using the "Add folder to Workspace" option. By doing so, we do not receive diagnostics, and we receive snippet suggestions.

**Comment by TrevCraw** [comment]
@anusreelakshmi934 Can you show the steps to go from a broken project to a working one using the "Add folder to Workspace" option? What does the project/workspace structure look like before and after? Can both projects be added to the same Workspace? And then can both projects be open in VS Code at the same time if they are part of the same Workspace?

On a separate note, what do you mean by "we do not receive diagnostics"?

**Comment by anusreelakshmi934** [comment]
When using the "Add folder to Workspace" feature, snippet suggestions become available when opening two files with the same app name. Here's how it works: 
1. First, open one of the files and check for snippet suggestions.
2. Next, go to File -> Add folder to Workspace, and select the other file with the same app name.
3. Now, check for snippet suggestions again, and you will receive them.

Now the workspace is UNTITLED (WORKSPACE). Even after saving the workspace, suggestions continue to be available. Both projects can be added to the same workspace and opened.

Regarding diagnostics, when I mentioned "We do not receive diagnostics," I meant that the diagnostic message "filename.java is a non-project file, only syntax errors are reported" is not displayed when adding projects using the "Add folder to Workspace" feature.

**Comment by anusreelakshmi934** [comment]
I'm currently facing difficulties with the workaround; I'm not receiving suggestions as expected. Additionally, I occasionally encounter a "unit is null" error when trying to open a Java file after adding the project using "Add folder to Workspace". This pertains to Issue #326 on GitHub for further context.  

Also I am encountering #312 while trying to start the language server after few attempts.

**Comment by anusreelakshmi934** [comment]
Still the workaround doesnt seem to be working. May need to further investigate some other ways for the issue. 
Still encountering [#312 ](https://github.com/OpenLiberty/liberty-tools-vscode/issues/312 )while trying to start the language server after few attempts.

**Comment by gilbysunil14** [comment]
Looked for multiple solutions. But as of now, cleaning the Java language server workspace, adding folders to the Java source path, setting sourcePaths in the the settings.json doesn’t work. Also adding projects individually to the workspace does not fix the problem.
What I am able to understand is that the vs code is not recognising the project as a java project.
https://git.eclipse.org/r/plugins/gitiles/jdt/eclipse.jdt.core/+/BETA_JAVA13/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModelManager.java#2524

I was also able to find similar issues reported in the redhat language support for java repo: https://github.com/redhat-developer/vscode-java/issues/2545. After duplicating this issue it was evident that there is no issue while having multiple projects in same workspace with different names. But the issue exists for multiple projects with the same name.

I checked to see if the same issue exists in other extensions. For this I tried with Project Manager for Java by Microsoft. There also same kind of issue exists where the extension does not identify multiple projects with the same name. See the below screenshot. There are two projects with name app-name-samename  imported to the workspace. But only one of them is recognised as the java project.
<img width="1728" alt="Screenshot 2024-11-04 at 6 54 08 PM" src="https://github.com/user-attachments/assets/af08893e-6574-41de-ab36-a8665e479683">

Do the following steps as workaround for this issue:

1. For Maven, use the `<finalName>` property under the `<build>` section of the pom.xml file or change the `<artifactId>`. For Gradle, use the `rootProject.name` property in settings.gradle file.
2. Press `Ctrl + Shift + P` (Windows/Linux) or `Cmd + Shift + P` (macOS) to open the Command Palette and run the command `Developer: Reload Window`
